oak-core/src/main/java/org/apache/jackrabbit/oak/core/MutableRoot.java

```diff
@@ -121,6 +126,8 @@ class MutableRoot implements Root, PermissionAware {
      */
     private final MoveTracker moveTracker = new MoveTracker();
 
+    private final Map<MutableTree, Object> trees = synchronizedMap(new WeakHashMap<>());
```

Comment:
Shouldn't we use weakSet here, since we are not using `value` from `Map`
```suggestion
    private final Set<MutableTree> treeSet = newSetFromMap(new WeakHashMap<MutableTree, Boolean>());
```